### DISCUSSION TITLE
Does uno support x:Bind method invocation for private members ?

### DISCUSSION THREAD

**Discussion by jeromelaban** [unoplatform/uno#9528] [Q&A]
[_Question from Emma111#5878 on Discord:_](https://discord.com/channels/372137812037730304/550416151172087808/1009372525765083196)

Hi, not sure this quite fits the github comment chain, a few questions

Does Uno support:

```xml
Child="{x:Bind m_viewFactory.CreateView(m_viewModel.PaneViewType), Mode=OneWay}"
```

where `m_viewFactory` and `m_viewModel` are private members of the `UserControl`?

Debugger shows that Uno does register a handler for `m_viewModel.PaneViewType` and it is invoked. However, `CreateView` is never called from Uno.

Currently isolating this section into it's own small test project

**Comment by jeromelaban** [comment]
Private members binding is supported, as well as invocation in methods, but there is a difference with the way WinUI works, seen in https://github.com/unoplatform/uno/issues/2872. It may be that you're setting `m_viewModel` too late for the changes to be picked up the Uno's binding engine.

**Reply by Tapir28** [reply]
```
ViewType m_paneViewType;
ViewType IDevicesPageVm.PaneViewType => m_paneViewType;
ViewType PaneViewType { set => SetAndNotify(ref m_paneViewType, value); }
```

Under debugger I can see that the PropertyChanged event for PaneViewType is actually triggered. After setting a breakpoint on line 25 of DevicesPageVm.cs, I can single step into SetAndNotify and see that PropertyChanged event is invoked and that it is not null. Debugger shows that a delegate from Uno is added to the event, and it is called.

But this call does not result in the line 24 being called, which would occur when Uno is getting the actual value of PaneViewType as a result of their PropertyChanged handler running. Neither m_viewFactory.CreateView() is called. As if the Uno handler is invoked but does nothing.

Is it possible to step into the SourceCode of Uno when the PropertyChanged handler is called? Currently it's non-debuggable.

**Reply by jeromelaban** [reply]
Thanks. Would you be able to create a standalone repro of this behavior? Also, can you try doing something like this:

```xaml
<StackPanel>
   <TextBlock Text="{x:Bind m_viewFactory, Mode=OneWay}" />
   <TextBlock Text="{x:Bind m_viewModel.PaneViewType, Mode=OneWay}" />
</StackPanel>
```

And see if what you get in the TextBlocks, visually?

**Reply by Rybasum** [reply]
@jeromelaban Does it matter that the getter of PaneViewType is implemented explicitly (through IDevicesPageVm) rather that being public? I remember that this used to be an issue in early UWP until Microsoft fixed it...

**Reply by jeromelaban** [reply]
@Rybasum it's not supposed to be an issue. Let me know what you see with the two TextBlock above, it may help find the root cause (aside from an actual standalone repro).

**Reply by jeromelaban** [reply]
The issue here is access to members on the bound types. The current implementation of x:Bind does not allow for binding on members this way, and requires members to be public on the actual instance being bound.

The workaround here is to do this:
```
public ViewType PaneViewType { set => SetAndNotify(ref m_paneViewType, value); get => m_paneViewType; }
```
to both have a getter and have `PaneViewType` being public.

Let's track this in https://github.com/unoplatform/uno/issues/9539

**Reply by Tapir28** [reply]
This solved the issue, thank you!

**Comment by Tapir28** [comment]
Yep, I'll put it in a Repo, thanks!

**Reply by Tapir28** [reply]
https://github.com/Tapir28/TestMacFunctionBinding

Standalone repo of the issue.

On windows, Both textboxes work, as well as the  ViewFactory.
On Mac however, none of them seem to.
[TestMacFunctionBinding.zip](https://github.com/unoplatform/uno/files/9371871/TestMacFunctionBinding.zip)

**Reply by Tapir28** [reply]
Just realised I replied to myself, so u might not have gotten a notification @jeromelaban apologies if you did already recieve one.
